### PR TITLE
autodoc: Don't rely on broken heuristic

### PR DESCRIPTION
**Pull request by khwilliamson**
Prior to this commit, the code was looking at the name of the element to decide if a pTHX parameter should be displayed.  It would display it if the name started with "Perl_".  That mostly worked, but it could be wrong, and was in two cases: for Perl_custom_op_xop() and Perl_localeconv().

The logic was complicated and existed in an earlier loop, and was not repeated in the loop doing the pTHX calculation.

But it turns out that by just moving that calculation to the earlier loop, the logic doesn't have to be repeated.

### CODE REVIEW

File: autodoc.pl [hunk 1]
```diff
@@ -1398,21 +1398,43 @@ ($$$)
...
-        if (   ($item_flags =~ /p/ && $item_flags =~ /o/ && $item_flags !~ /M/)
 
-                # Can't handle threaded varargs
-            || (   $item_flags =~ /f/
-                && $item_flags !~ /T/
-                && $item_name !~ /strftime/))
-        {
+        # If only the Perl_foo form is to be displayed, change the name of
+        # this item to be that.  This happens for either of two reasons:
+        #   1) The flags say we want "Perl_", but also to not create an entry
```
Comment (jkeenan) [inline autodoc.pl:1404]: `but also to not create an entry ...` -- confused as to what that's supposed to mean.  Did you intend `do not create`?  Or perhaps, `but we also do not want to create an entry ...`?

File: autodoc.pl [hunk 2]
```diff
@@ -1431,18 +1453,11 @@ ($$$)
...
-                                            ? "pTHX_"
-                                            : "pTHX")
-                                                   if $item->{flags} !~ /T/
-                                                   && $item->{name} =~ /^Perl_/;
-            }
-
-            # Look through all the items in this entry.  Find the longest
-            # return type and longest name so that if multiple ones are shown,
-            # they can be nicely vertically aligned.
+            # Look through all the items in this entry.  If all have the same
```
Comment (jkeenan) [inline autodoc.pl]: suggest: `s/If all have/If they all have/`

File: autodoc.pl [hunk 3]
```diff
@@ -1431,18 +1453,11 @@ ($$$)
...
+            # return type and arguments (including thread context), only the
+            # main entry is displayed.
+            # Also, find the longest return type and longest name so that if
+            # multiple ones are shown, they can be vertically aligned nicely
```
Comment (jkeenan) [inline autodoc.pl]: Suggest: period after `nicely`

**Comment by jkeenan** [comment]
I built blead through `make test_prep`, then did likewise for the branch in this p.r. rebased on blead.  When I examined the respective `pod/perlapi.pod` generated files, the only differences I found were these:
```
$ diff -w ../perl/pod/perlapi.pod ./pod/perlapi.pod
5612c5612
<  const XOP *  Perl_custom_op_xop(pTHX_ const OP *o)
---
>  const XOP *  Perl_custom_op_xop(const OP *o)
16727c16727
<  HV *  Perl_localeconv(pTHX)
---
>  HV *  Perl_localeconv()
```
Is that what you were expecting?  Thanks.

File: autodoc.pl (hunk 1 above)
Comment (tonycoz) [reply autodoc.pl:1404]: Normally we define functions in the API with a `Perl_` prefix, *and* define a macro without the prefix to call that function, adding a thread context parameter if needed, eg.
```
/* proto.h */
PERL_CALLCONV void
Perl_sv_setpvn(pTHX_ SV * const sv, const char * const ptr, const STRLEN len);

/* embed.h */
# define sv_setpvn(a,b,c)                       Perl_sv_setpvn(aTHX_ a,b,c)
```
In some cases we have the `Perl_` prefix but don't want the macro to be defined.

Note that the code being modified here doesn't define the macro, it's just documenting how to call the function, so "to not create" is more reasonable than "do not create".

**Review by tonycoz — APPROVED** [review]
Looks fine except for @jkeenan 's comment nits.

File: autodoc.pl (hunk 2 above)
Comment (khwilliamson) [reply autodoc.pl]: done

File: autodoc.pl (hunk 3 above)
Comment (khwilliamson) [reply autodoc.pl]: done

File: autodoc.pl (hunk 1 above)
Comment (khwilliamson) [reply autodoc.pl:1404]: Unchanged

**Comment by khwilliamson** [comment]
This is what I was expecting; the commit message mentions both cases, and that there are none other

**Pull request merged**